### ISSUE TITLE
map checks dimensions inconsistently

### ISSUE THREAD

**Issue by cstjean** [JuliaLang/julia#29523]
I believe that the second example should also be a `DimensionMismatch`, for consistency.
```julia
julia> map(<, [1, 2], [2])
ERROR: DimensionMismatch("dimensions must match")
Stacktrace:
 [1] promote_shape at ./indices.jl:154 [inlined]
 [2] axes(::Base.Iterators.Zip2{Array{Int64,1},Array{Int64,1}}) at ./iterators.jl:291
 [3] _array_for at ./array.jl:611 [inlined]
 [4] collect(::Base.Generator{Base.Iterators.Zip2{Array{Int64,1},Array{Int64,1}},getfield(Base, Symbol("##3#4")){typeof(<)}}) at ./array.jl:624
 [5] map(::Function, ::Array{Int64,1}, ::Array{Int64,1}) at ./abstractarray.jl:2060
 [6] top-level scope at none:0

julia> map(<, [1 2], [2])
1-element Array{Bool,1}:
 true
```

**Comment by blegat** [comment]
I have extracted a minimal failing example from the bug in Polyhedra. It does not require any packages:
```julia
abstract type A{T} end
abstract type B{T} end
const C{T} = Union{B{T}, A{T}}
struct MyVec <: B{Int}
    elements::Vector{Int}
end
function startindex(elements::Vector)
    if isempty(elements)
        return nothing
    else
        return 1
    end
end
struct Iterator{PT<:Tuple{C}}
    ps::PT
    function Iterator(ps::PT) where {PT}
        new{PT}(ps)
    end
end
element_and_index(it, idx::Nothing) = nothing
element_and_index(it::Iterator, idx::Int) = it.ps[1].elements[idx], idx
function Base.iterate(it::Iterator{Tuple{T}}) where {T}
    idx = startindex(it.ps[1].elements)
    return element_and_index(it, idx)
end

h = MyVec([1])
let
    i = 0
    for a in Iterator((h,))
        i += 1 # never gets executed
    end
    println(i) # Prints 0 instead of 1
end
```
It seems that every element is important in this example. Whenever I try to simplify it, the bug disappears :-P

**Comment by BioTurboNick** [comment]
My issue #33635 is related although different. Scalar inputs are treated like iterables with 1 element, and documents other potentially inconsistent behavior with e.g. generators and tuples as arguments.

As a user, I would wish map()/pmap() to enforce size wherever possible. If a user desires the iterable-as-many-as-possible behavior, I'd argue they should be explicit about that.

**Comment by mschauer** [comment]
Resolved by #29927
